apps/system/js/scheduler.js

```diff
@@ -0,0 +1,64 @@
+/* global BaseModule */
+'use strict';
+
+(function() {
+  var Scheduler = function() {
+  };
+  Scheduler.EVENTS = [
+    'homescreenloaded'
+  ];
+  Scheduler.SERVICES = [
+    'schedule'
+  ];
+  BaseModule.create(Scheduler, {
+    name: 'Scheduler',
+    DEBUG: false,
+    _start: function() {
+      this._queue = [];
+      if (!this.service.query('getHomescreen')) {
+        this._loadingHomescreenFirstTime = true;
+      }
+    },
+    _handle_homescreenloaded: function() {
+      window.removeEventListener('homescreenloaded', this);
+      if (this._loadingHomescreenFirstTime) {
+        document.body.setAttribute('ready-state', 'visuallyLoaded');
+        window.performance.mark('visuallyLoaded');
+      }
+      this.execute();
+    },
+    execute: function() {
+      var length = this._queue.length;
+      this.debug('process operations in the schedule pool: ' + length);
+      this._queue.forEach(function(operation) {
+        operation.request();
+        operation.resolve();
+      });
+      this._queue = [];
+      if (length > 0 && this._loadingHomescreenFirstTime) {
+        this._loadingHomescreenFirstTime = false;
+        // XXX: Actually we need to have a module list
+        //      to know the actual loaded for all modules.
+        this.debug('fully loaded');
+        document.body.setAttribute('ready-state', 'fullyLoaded');
+        window.performance.mark('fullyLoaded');
+      }
+    },
+    schedule: function(request) {
```

Comment:
Having this method take a callback seems very strange to me, and seems to defeat the purpose of a promise. Why does it need a callback?